File: python-prototype/soundplace/onset.py
"""Onset 检测 (对应 src-tauri/src/analysis/onset.rs).

Spectral Flux + HFC 组合, 自适应阈值 (median + k×MAD), 去抖.

优化 (针对实时捕获背景噪声过触发):
- 增加 min_absolute_flux 阈值: 背景噪声 flux < 1.0 时直接拒绝
- 提高 threshold_k 从 3.0 → 5.0 (更严格)
- 增加最小间隔 100ms → 150ms
- intensity 归一化用 max(threshold, min_absolute_flux) 避免除以小数放大噪声
"""

from __future__ import annotations

import time
from collections import deque
from typing import Optional

import numpy as np

# Spectrum 类型别名 (与 fft.py 一致)
Spectrum = np.ndarray

# 背景噪声的典型 flux 水平 (经验值, 低于此值认为是噪声)
# 实测: 静音时 flux ≈ 0, 系统空闲时 flux ≈ 0.1-0.5, 音乐瞬时 flux > 5
# 调低一点 (从 1.0 → 0.5) 以便真实声音能触发
MIN_ABSOLUTE_FLUX: float = 0.5


class OnsetEvent:
    """Onset 检测事件."""

    __slots__ = ("intensity", "timestamp_ms")

    def __init__(self, intensity: float, timestamp_ms: int) -> None:
        self.intensity = float(intensity)
        self.timestamp_ms = int(timestamp_ms)


class OnsetDetector:
    """Onset 检测器."""

    def __init__(
        self,
        window_size: int = 30,
        threshold_k: float = 5.0,
        min_interval_ms: int = 150,
        min_absolute_flux: float = MIN_ABSOLUTE_FLUX,
    ) -> None:
        self.prev_spectrum: Optional[np.ndarray] = None
        self.flux_history: deque[float] = deque(maxlen=window_size)
        self.window_size = window_size
        self.threshold_k = threshold_k
        self.last_trigger_ts: int = 0
        self.min_interval_ms = min_interval_ms
        self.min_absolute_flux = min_absolute_flux

    def detect(self, spectrum: np.ndarray) -> Optional[OnsetEvent]:
        """检测当前帧是否触发 onset.

        Args:
            spectrum: FFT 幅度谱 (1D float32 数组).

        Returns:
            触发时返回 OnsetEvent, 否则 None.
        """
        now_ms = int(time.time() * 1000)

        # 1. Spectral Flux: 正频谱差之和
        if self.prev_spectrum is not None and len(self.prev_spectrum) == len(spectrum):
            diff = spectrum - self.prev_spectrum
            flux = float(np.sum(np.maximum(diff, 0.0)))
        else:
            flux = 0.0

        # 2. HFC (归一化): 高频加权能量
        k_axis = np.arange(len(spectrum), dtype=np.float32)
        hfc = float(np.sum(k_axis * spectrum) / max(len(spectrum), 1))

        # 3. 组合特征: flux + 0.3 * hfc
        combined = flux + 0.3 * hfc

        # 4. 更新历史
        self.flux_history.append(combined)
        self.prev_spectrum = spectrum.copy()

        # 历史不足, 无法判断阈值
        if len(self.flux_history) < 10:
            return None

        # 5. 自适应阈值: median + k * MAD
        threshold = self._adaptive_threshold()
        # 取自适应阈值和绝对阈值中的较大者
        effective_threshold = max(threshold, self.min_absolute_flux)
        if combined <= effective_threshold:
            return None

        # 6. 去抖
        if now_ms - self.last_trigger_ts < self.min_interval_ms:
            return None

        # 7. 计算强度 (归一化 [0, 1])
        # 用 effective_threshold 做分母, 避免 background 噪声放大
        intensity = max(0.0, min(1.0, (combined - effective_threshold) / (effective_threshold + 1e-6)))
        self.last_trigger_ts = now_ms
        return OnsetEvent(intensity=intensity, timestamp_ms=now_ms)

    def _adaptive_threshold(self) -> float:
        """median + k * MAD."""
        sorted_vals = np.sort(np.array(self.flux_history, dtype=np.float32))
        n = len(sorted_vals)
        mid = n // 2
        if n % 2 == 0:
            median = float((sorted_vals[mid - 1] + sorted_vals[mid]) / 2.0)
        else:
            median = float(sorted_vals[mid])

        deviations = np.sort(np.abs(sorted_vals - median))
        if n % 2 == 0:
            mad = float((deviations[mid - 1] + deviations[mid]) / 2.0)
        else:
            mad = float(deviations[mid])

        return median + self.threshold_k * mad

File: src-tauri/src/analysis/onset.rs
// 瞬态事件检测 (Onset Detection)
//
// 基于 Spectral Flux + HFC 两个特征检测瞬态事件(脚步/枪声)
//
// 算法:
// 1. Spectral Flux: 正频谱差之和, 反映能量上升
//    flux = Σ max(0, |X_t[k]| - |X_{t-1}[k]|)
// 2. HFC (High Frequency Content): 高频能量加权, 对打击声敏感
//    hfc = Σ k * |X_t[k]|
// 3. 自适应阈值: 滑动窗口(约 20 帧)计算中位数 + k×MAD
// 4. 去抖: 同一类型事件最小间隔 100ms (可由 profile 配置)

use std::collections::VecDeque;
use std::time::{SystemTime, UNIX_EPOCH};

use super::fft::Spectrum;

/// Onset 检测事件
#[derive(Clone, Debug)]
pub struct OnsetEvent {
    /// 强度 [0.0, 1.0]
    pub intensity: f32,
    /// 时间戳(毫秒)
    pub timestamp: u64,
}

/// Onset 检测器
pub struct OnsetDetector {
    /// 上一帧的频谱(用于计算 Spectral Flux)
    prev_spectrum: Option<Spectrum>,
    /// 滑动窗口存储最近 N 帧的 flux 值, 用于自适应阈值
    flux_history: VecDeque<f32>,
    /// 滑动窗口大小
    window_size: usize,
    /// 阈值倍数 k (阈值 = median + k * MAD)
    threshold_k: f32,
    /// 上次触发时间戳(毫秒), 用于去抖
    last_trigger_ts: u64,
    /// 最小事件间隔(毫秒)
    min_interval_ms: u64,
}

impl OnsetDetector {
    /// 创建默认配置的检测器
    pub fn new() -> Self {
        Self {
            prev_spectrum: None,
            flux_history: VecDeque::with_capacity(32),
            window_size: 20,
            threshold_k: 3.0,
            last_trigger_ts: 0,
            min_interval_ms: 100,
        }
    }

    /// 用 profile 参数构造
    pub fn with_config(threshold_k: f32, min_interval_ms: u32) -> Self {
        let mut d = Self::new();
        d.threshold_k = threshold_k;
        d.min_interval_ms = min_interval_ms as u64;
        d
    }

    /// 检测是否触发 onset
    ///
    /// 输入当前帧的频谱(左右声道任一, 通常用左声道或合并)
    pub fn detect(&mut self, spectrum: &Spectrum) -> Option<OnsetEvent> {
        let now_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_millis() as u64)
            .unwrap_or(0);

        // 计算 Spectral Flux
        let flux = match &self.prev_spectrum {
            Some(prev) if prev.len() == spectrum.len() => {
                let mut sum = 0.0f32;
                for (cur, p) in spectrum.iter().zip(prev.iter()) {
                    let diff = cur - p;
                    if diff > 0.0 {
                        sum += diff;
                    }
                }
                sum
            }
            _ => 0.0,
        };

        // 计算 HFC (归一化)
        let hfc: f32 = spectrum
            .iter()
            .enumerate()
            .map(|(k, mag)| k as f32 * mag)
            .sum::<f32>()
            / spectrum.len().max(1) as f32;

        // 组合特征: flux + 0.3 * hfc (flux 为主, hfc 辅助)
        let combined = flux + 0.3 * hfc;

        // 更新历史
        self.flux_history.push_back(combined);
        if self.flux_history.len() > self.window_size {
            self.flux_history.pop_front();
        }

        // 更新 prev_spectrum
        self.prev_spectrum = Some(spectrum.clone());

        // 历史不足时无法判断阈值
        if self.flux_history.len() < 5 {
            return None;
        }

        // 计算自适应阈值: median + k * MAD
        let threshold = self.adaptive_threshold();
        if combined <= threshold {
            return None;
        }

        // 去抖
        if now_ms.saturating_sub(self.last_trigger_ts) < self.min_interval_ms {
            return None;
        }

        // 计算强度(归一化到 [0, 1])
        let intensity = ((combined - threshold) / (threshold + 1e-6)).clamp(0.0, 1.0);

        self.last_trigger_ts = now_ms;
        Some(OnsetEvent {
            intensity,
            timestamp: now_ms,
        })
    }

    /// 计算自适应阈值 = median + k * MAD
    fn adaptive_threshold(&self) -> f32 {
        let mut sorted: Vec<f32> = self.flux_history.iter().copied().collect();
        sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

        let mid = sorted.len() / 2;
        let median = if sorted.len() % 2 == 0 {
            (sorted[mid - 1] + sorted[mid]) / 2.0
        } else {
            sorted[mid]
        };

        // MAD = Median Absolute Deviation
        let mut deviations: Vec<f32> = sorted.iter().map(|x| (x - median).abs()).collect();
        deviations.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let mad_mid = deviations.len() / 2;
        let mad = if deviations.len() % 2 == 0 {
            (deviations[mad_mid - 1] + deviations[mad_mid]) / 2.0
        } else {
            deviations[mad_mid]
        };

        median + self.threshold_k * mad
    }
}

impl Default for OnsetDetector {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_no_onset_for_silence() {
        let mut detector = OnsetDetector::new();
        let spectrum = vec![0.0; 1024];
        // 静音不应触发
        for _ in 0..30 {
            assert!(detector.detect(&spectrum).is_none());
        }
    }

    #[test]
    fn test_detect_onset_on_energy_burst() {
        let mut detector = OnsetDetector::new();
        let silence = vec![0.0; 1024];

        // 先喂 20 帧静音建立基线
        for _ in 0..20 {
            detector.detect(&silence);
        }

        // 突然能量爆发
        let loud: Spectrum = (0..1024).map(|i| (i as f32 / 100.0).sin() * 10.0).collect();
        let event = detector.detect(&loud);
        assert!(event.is_some(), "应在能量爆发时触发 onset");
    }
}
